---
title: "Target Illumination GWAS Analytics (TIGA)"
author: "Jeremy Yang"
output:
  html_document:
    number_sections: yes
    toc: yes
  pdf_document:
    toc: yes
editor_options:
  chunk_output_type: console
---

# Introduction

Genome wide association studies (GWAS) can reveal important genotype to phenotype
associations, however, data quality and interpretability issues must be addressed.
The TIGA approach enables rational ranking, filtering and interpretation of GWAS
via metrics, methods, and interactive visualization.  Each inferred gene-trait
association is evaluated for confidence, with scores solely derived
from aggregated statistics, linking a protein-coding gene and phenotype.

* Data source: [GWAS Catalog](https://www.ebi.ac.uk/gwas/).
* Workflow designed for integration with TCRD & Pharos
* RCRAS, Relative Citation Ratio (RCR) Aggregated Score based on iCite-RCR.
* Confidence higher (and provenance/accountability) for `mapped` over `reported` SNP-gene associations.
* BETA not comparable for different units, thus using N_beta counts for TIGA.
* RShiny app: <http://unmtid-shinyapps.net/tiga/>

## About NHGRI-EBI GWAS Catalog

GWAS Catalog (<http://www.ebi.ac.uk/gwas/>) studies each have a `study_accession`.
Also are associated with a publication (PubMedID), but not uniquely.
See <https://www.ebi.ac.uk/gwas/docs/fileheaders>.

Key definitions:

	`MAPPED GENE(S)`: Gene(s) mapped to the strongest SNP. If the SNP is located
	within a gene, that gene is listed. If the SNP is intergenic, the upstream
	and downstream genes are listed, separated by a hyphen.
	`UPSTREAM_GENE_ID`: Entrez Gene ID for nearest upstream gene to rs number, if not within gene
  `DOWNSTREAM_GENE_ID`: Entrez Gene ID for nearest downstream gene to rs number, if not within gene
  `SNP_GENE_IDS`: Entrez Gene ID, if rs number within gene; multiple genes denotes overlapping transcripts
  `UPSTREAM_GENE_DISTANCE`: distance in kb for nearest upstream gene to rs number, if not within gene
  `DOWNSTREAM_GENE_DISTANCE`: distance in kb for nearest downstream gene to rs number, if not within gene
	`OR or BETA`: Reported odds ratio or beta-coefficient associated with
	strongest SNP risk allele. Note that if an OR <1 is reported this is
	inverted, along with the reported allele, so that all ORs included in
	the Catalog are >1. Appropriate unit and increase/decrease are included
	for beta coefficients.
	`INITIAL SAMPLE DESCRIPTION`: Sample size and ancestry description for
	stage 1 of GWAS (summing across multiple Stage 1 populations, if applicable)
	`GENOTYPING_TECHNOLOGY`: Genotyping technology/ies used in this study, with
	additional array information (ex. Immunochip or Exome array) in brackets.

*  Buniello A, MacArthur JAL, Cerezo M, Harris LW, Hayhurst J, Malangone C, et al. The NHGRI-EBI GWAS Catalog of published genome-wide association studies, targeted arrays and summary statistics 2019. Nucleic Acids Res. 2019;47: D1005â€“D1012.
 

```{r echo=FALSE}
knitr::opts_chunk$set(echo=FALSE, message=FALSE)
base::date()
```

```{r message=FALSE, warning=FALSE}
library(readr, quietly=T)
library(data.table, quietly=T)
library(plotly, quietly=T)
```

# Read files

From GWAS Catalog, TCRD, and EFO.

```{r}
GC_REL <- trimws(read_file("LATEST_RELEASE.txt"))
message(sprintf("LATEST_RELEASE: %s", GC_REL))
```

```{r}
ODIR <- sprintf("data/%s", gsub("\\-", "", GC_REL))
ifile_gwas <- paste0(ODIR, "/gwascat_gwas.tsv")
ifile_assn <- paste0(ODIR, "/gwascat_assn.tsv")
ifile_snp2gene <- paste0(ODIR, "/gwascat_snp2gene_MERGED.tsv")
ifile_trait <- paste0(ODIR, "/gwascat_trait.tsv")
ifile_icite <-paste0(ODIR, "/gwascat_icite.tsv")
ifile_tcrd <- paste0(ODIR, "/tcrd_targets.tsv")
ifile_efo <- paste0(ODIR, "/efo.tsv")
ifile_efo_groups <- paste0(ODIR, "/efo_groups.tsv")
ifile_efo_gwas <- paste0(ODIR, "/efo_sub_gwas.tsv")
gt_provfile <- paste0(ODIR, "/gt_provenance.tsv.gz")
gt_statsfile <- paste0(ODIR, "/gt_stats.tsv.gz")
gwas <- read_delim(ifile_gwas, "\t", col_types=cols(.default=col_character(), 
    DATE=col_date(format="%Y-%m-%d"), DATE_ADDED_TO_CATALOG=col_date(format="%Y-%m-%d"),
    ASSOCIATION_COUNT=col_integer(), study_N=col_integer()))
setDT(gwas)
assn <- read_delim(ifile_assn, "\t", col_types=cols(.default=col_character(), INTERGENIC=col_logical(), `P-VALUE`=col_double(), PVALUE_MLOG=col_double(), OR_or_BETA=col_double(), UPSTREAM_GENE_DISTANCE=col_integer(), DOWNSTREAM_GENE_DISTANCE=col_integer(), DATE=col_date(format="%Y-%m-%d"), DATE_ADDED_TO_CATALOG=col_date(format="%Y-%m-%d")))
setDT(assn)
snp2gene <- read_delim(ifile_snp2gene, "\t", col_types=cols(.default=col_character(),
    MAPPED_OR_REPORTED=col_factor(c("r", "m", "md", "mu"))))
setDT(snp2gene)
trait <- read_delim(ifile_trait, "\t", col_types=cols(.default=col_character()))
setDT(trait)
tcrd <- read_delim(ifile_tcrd, "\t", col_types=cols(.default=col_character(), idgList=col_logical()))
setDT(tcrd)
```

# Genome Wide Association Studies (GWAS)

### Counts by year, platform

Grouped by vendor (first in list if multiple), though technologies may have evolved for a given
vendor.

```{r}
sprintf("Studies: %d; traits: %d ; PMIDs: %d", uniqueN(gwas$STUDY_ACCESSION), uniqueN(gwas$DISEASE_TRAIT), uniqueN(gwas$PUBMEDID))
gwas[, `:=`(YEAR = format(DATE,"%Y"), PLATFORM = sub("[, ].*$", "", `PLATFORM_[SNPS_PASSING_QC]`))]
gwas[, YEAR := as.integer(YEAR)]
gwas_platforms <- gwas[, .(.N), by=c("YEAR", "PLATFORM")]
setorder(gwas_platforms, YEAR, PLATFORM)
gwas_platforms_cast <- dcast(gwas_platforms, YEAR ~ PLATFORM, value.var="N", fill=0)
plot_ly(gwas_platforms_cast, x=~YEAR, y=~NR, name="Not Reported", type="scatter", mode="markers", fill="tonexty") %>%
  add_trace(y = ~Affymetrix, name="Affymetrix") %>%
  add_trace(y = ~Illumina, name="Illumina") %>%
  layout(title = 'GWAS counts by year &amp; vendor',
         xaxis = list(title="", showgrid=F, range=c(2012, 2021), dtick=1, tickangle=45),
         yaxis = list(title="", showgrid=F),
         margin = list(t=100, l=100),
         font = list(family="monospace", size=24),
         legend = list(x=.1, y=.7), showlegend=T) %>%
  add_annotations(text=("N = count(studies)"), showarrow=F, x=0.05, y=1, xref="paper", yref="paper")
```

### Sample size

Statistical power and strength of evidence is dependent on size of the cohort.
Sample size box plots by year.

```{r}
plot_ly(gwas[YEAR>2010 & YEAR<2022], x=~YEAR, y=~study_N, type="box") %>%
    layout(title="GWAS sample sizes",
         xaxis=list(title="", tickangle=45), yaxis=list(title="Study_N", type="log"),
         margin = list(t=100, l=100),
         font = list(family="monospace", size=24)) %>%
  add_annotations(text=sprintf("median:%d  mean:%d  range:[%d-%.2g]", median(gwas$study_N, na.rm=T), round(mean(gwas$study_N, na.rm=T)), min(gwas$study_N, na.rm=T), max(gwas$study_N, na.rm=T)), showarrow=F, x=0.5, y=1, xref="paper", yref="paper")
```


```{r}
bins <- c(0, 5e2, 1e3, 1e4, 5e4, 1e5, Inf)
minN = bins[1:length(bins)-1]
maxN = bins[2:length(bins)]
gwas_counts <- data.table(minN=minN, maxN=maxN, count = NA)
for (i in 1:nrow(gwas_counts)) {
  gwas_counts$count[i] <- uniqueN(gwas[(study_N>=minN[i] & study_N<maxN[i]), STUDY_ACCESSION])
}
gwas_counts[, name := sprintf("(%6g <= N < %6g)", minN, maxN)]
gwas_counts[, pct := 100 * count / sum(count)]
write_delim(gwas_counts, "data/gwas_counts_by_samplesize.tsv", "\t")
knitr::kable(gwas_counts[, .(Sample_size = name, Study_count = count, Pct = pct)], caption="GWAS sample sizes", digits=1)
```

### Journals

```{r}
gwas_journals <- gwas[, .(N_gwas = .N, N_pmid = uniqueN(PUBMEDID), N_assn = sum(ASSOCIATION_COUNT)), by="JOURNAL"]
setorder(gwas_journals, -N_gwas)
write_delim(gwas_journals, "data/gwas_counts_by_journal.tsv", "\t")
knitr::kable(gwas_journals[1:12], caption="Top journals by N_gwas")
```

# Relative Citation Ratio (RCR)

RCR annotations from iCite API; PMIDs from GWASCatalog.
We estimate RCR=median as reasonable prior for new publications and any lacking RCR from iCite. But don't include in 
visualization since these will obscure distribution.

```{r warning=FALSE}
icite <- read_delim(ifile_icite, "\t", col_types=cols(.default=col_character(),
                       relative_citation_ratio=col_double(), field_citation_rate=col_double(),
                       citation_count=col_integer(), nih_percentile=col_double(),
                       expected_citations_per_year=col_double(), citations_per_year=col_double(),
                       year=col_integer()))
gwas_icite <- merge(gwas[, .(STUDY_ACCESSION, PUBMEDID)], icite, by.x="PUBMEDID", by.y="pmid", all.x=T, all.y=F)
qtl <- quantile(gwas_icite$relative_citation_ratio, seq(0, 1, .1), na.rm=T)
rcr_max <- 15
year_min <- min(gwas_icite[, year], na.rm=T)
year_max <- as.integer(format(Sys.Date(), "%Y"))-2
anns <- c(sprintf("N_pmid = %d", nrow(gwas_icite)), sprintf("mean = %.1f ; median = %.1f ; max = %.1f", mean(gwas_icite$relative_citation_ratio, na.rm=T), median(gwas_icite$relative_citation_ratio, na.rm=T), max(gwas_icite$relative_citation_ratio, na.rm=T)), sprintf("90%%ile = %.1f", qtl["90%"]), sprintf("%d - %d", year_min, year_max),
    sprintf("Plot RCR <= %d; count(RCR > %d) = %d", rcr_max, rcr_max, gwas_icite[relative_citation_ratio>rcr_max, .N]))
plot_ly(gwas_icite[year<=year_max], x=~relative_citation_ratio, type="histogram", marker=list(color="#22AA22")) %>%
  layout(title = 'Relative Citation Ratio (RCR) distribution',
         xaxis = list(title="RCR", type="normal", range=c(0, rcr_max), showgrid=T),
         yaxis = list(title="", showgrid=T),
         margin = list(t=100, l=100, r=100),
         font = list(family="monospace", size=18)) %>%
  add_annotations(text=paste0(anns, collapse="<br>"), showarrow=F, x=0.5, y=.9, xref="paper", yref="paper")
```

### RCR for top GWAS journals by paper count

```{r warning=FALSE}
gwas_icite[!(journal %in% gwas_icite[, .N, by=journal][order(-N)][1:10, journal]), journal := "Other"]
wordwrap <- function(v, len) sapply(sapply(v, strwrap, len), paste, collapse="<br>")
plot_ly(gwas_icite, 
        x=~year, 
        y=~relative_citation_ratio, 
        color = ~journal, 
        type="box",
        text=sprintf("%s: %s, <i>%s</i> (%d)", gwas_icite$PUBMEDID, wordwrap(gwas_icite$title, 40), gwas_icite$journal, gwas_icite$year)) %>%
  layout(title = 'iCite Relative Citation Ratio<br>for top journals by GWAS publication count',
         xaxis = list(title="", range=c(2015.5, 2020.5), showgrid=F, dtick=1, tickangle=45, tickfont=list(size=14)),
         yaxis = list(title="RCR", range=c(0,25), showgrid=T),
         margin = list(t=100, l=100, r=100),
         font = list(family="monospace", size=18),
         boxmode="group",
         legend=list(x=1, y=.9, font=list(size=12)), showlegend=T)
```

# Associations (SNP to trait)

```{r}
sprintf("Associations total: %d ; SNPs: %d ; traits: %d ; PMIDs: %d", nrow(assn), uniqueN(assn$SNPS), uniqueN(assn$DISEASE_TRAIT), uniqueN(assn$PUBMEDID))
```

## CONTEXT aka functionalClass

```{r}
funClass <- assn[, .N, by="CONTEXT"]
funClass <- unique(funClass[grepl(";", CONTEXT), `:=`(CONTEXT = "Multiple", N = sum(N))])
setorder(funClass, -N)
write_delim(funClass, "data/snp_context_counts.tsv", "\t")
knitr::kable(funClass[N>=10], caption="SNP functional classes (aka CONTEXT), N>=10")
plot_ly(type="pie", values=funClass$N, labels=paste0(funClass$CONTEXT, " (", funClass$N, ")"), textinfo="label+percent", textfont=list(size=12) ) %>%
  layout(title="SNP functional classes (aka CONTEXT)",
         margin = list(t=100, b=100),
         font=list(family="monospace", size=18), showlegend=F)
```

## (UP|DOWN)STREAM_GENE_DISTANCE

```{r}
sprintf("Associations with UPSTREAM_GENE_DISTANCE: %d (%.1f%%)", assn[!is.na(UPSTREAM_GENE_DISTANCE), .N], 100*assn[!is.na(UPSTREAM_GENE_DISTANCE), .N]/nrow(assn))
sprintf("Associations with DOWNSTREAM_GENE_DISTANCE: %d (%.1f%%)", assn[!is.na(DOWNSTREAM_GENE_DISTANCE), .N], 100*assn[!is.na(DOWNSTREAM_GENE_DISTANCE), .N]/nrow(assn))
qtl <- quantile(assn$UPSTREAM_GENE_DISTANCE, na.rm=T)
sprintf("UPSTREAM: %4sile: %d\n", names(qtl), as.integer(qtl))
qtl <- quantile(assn$DOWNSTREAM_GENE_DISTANCE, na.rm=T)
sprintf("DOWNSTREAM: %4sile: %d\n", names(qtl), as.integer(qtl))
```


```{r}
xmax <- 1e5
xvals <- seq(0, xmax, by=1000)
annos1 <- c(
	sprintf("N_total = %d", assn[, .N]))
	#sprintf("N_[BP_up>%.1g] = %d", xmax, assn[UPSTREAM_GENE_DISTANCE>xmax, .N]),
	#sprintf("N_[BP_down>%.1g] = %d", xmax, assn[DOWNSTREAM_GENE_DISTANCE>xmax, .N]))
annos2 <- c(
  sprintf("Exponential weighting function"),
  sprintf("f(x) = 2<sup>(-x/50k)</sup>")
)
plots <- list(
  plot_ly(alpha=0.6) %>%
    add_histogram(name="UPSTREAM", x=assn[(INTERGENIC) & !is.na(UPSTREAM_GENE_DISTANCE) & UPSTREAM_GENE_DISTANCE<xmax, UPSTREAM_GENE_DISTANCE], histnorm="probability") %>%
    add_histogram(name="DOWNSTREAM", x=assn[(INTERGENIC) & !is.na(DOWNSTREAM_GENE_DISTANCE) & DOWNSTREAM_GENE_DISTANCE<xmax, DOWNSTREAM_GENE_DISTANCE], histnorm="probability") %>%
  add_annotations(text=paste0(annos1, collapse="<br>"), x=1, xanchor="right", y=.7, yanchor="bottom", xref="paper", yref="paper", showarrow=F) %>%
  add_annotations(text="Distribution of mapped genomic distances", x=.5, xanchor="center", y=.7, yanchor="top", xref="paper", yref="paper", showarrow=F) %>%
    layout(yaxis=list(title="probability")),
  plot_ly() %>%
    add_trace(name="f(x)", type="scatter", x=xvals, y=exp(-xvals/4e4*log(2)), mode="lines", line=list(width=5, color='#CC2222')) %>%
      add_annotations(text=paste0(annos2, collapse="<br>"), x=.5, xanchor="center", y=.7, yanchor="top", xref="paper", yref="paper", showarrow=F) %>%
    layout(yaxis=list(title="f(x)"))
  )
subplot(plots, nrows=2, shareX=T, titleX=F, titleY=T, margin=.05) %>%
  layout(title="GWAS MAPPED SNP-GENE DISTANCE",
         margin = list(l=80, t=100, r=50),
         xaxis=list(title="BPs", range=c(0, xmax)),
         font=list(family="monospace", size=24),
         legend = list(x=.8, y=.4), showlegend=T)
```

# SNP to gene mappings

* Catalog mapping via __Ensembl Pipeline__:
* `GSYMB`, `MAPPED_GENE` fields may include chromosomal locations or be "intergenic".
* Chromosomal locations (LOC*) may be non-coding RNA, etc.
* Mappings _within_ gene stronger associations than _upstream_ or _downstream_.
* We understand _upstream_ associations stronger than _downstream_ at equal distance.

```{r}
sprintf("snp2gene: total associations: %d ; studies: %d ; snps: %d ; genes: %d ; intergenic associations: %d ; chromosomal location associations: %d",
        nrow(snp2gene), uniqueN(snp2gene$STUDY_ACCESSION), uniqueN(snp2gene$SNP), 
        uniqueN(snp2gene[!grepl("^LOC", GSYMB) & GSYMB != "intergenic", .(GSYMB)]),
        nrow(snp2gene[GSYMB=="intergenic"]), nrow(snp2gene[grepl("^LOC", GSYMB)]))
snp2gene[, MAPPED_OR_REPORTED := sub('^r$', 'reported', MAPPED_OR_REPORTED)]
snp2gene[, MAPPED_OR_REPORTED := sub('^m$', 'mapped_within', MAPPED_OR_REPORTED)]
snp2gene[, MAPPED_OR_REPORTED := sub('^mu$', 'mapped_upstream', MAPPED_OR_REPORTED)]
snp2gene[, MAPPED_OR_REPORTED := sub('^md$', 'mapped_downstream', MAPPED_OR_REPORTED)]
snp2gene_counts <- snp2gene[, .(N_snp = .N), by="MAPPED_OR_REPORTED"]
knitr::kable(snp2gene_counts)
```
# Ensemble annotations

Using Ensembl IDs, we query the Ensembl API for additional annotations including gene ___biotype___, thereby filtering for __protein_coding__.

```{r}
ensbInfo <- read_delim("data/gwascat_EnsemblInfo.tsv.gz", "\t", col_types = cols(.default=col_character(), version=col_integer(), strand=col_integer(), start=col_integer(), end=col_integer()))
setDT(ensbInfo)
biotypes <- ensbInfo[, .(N_gene = .N), by="biotype"][order(-N_gene)]
biotypes <- biotypes[, Percent := 100 * N_gene / sum(N_gene)]
write_delim(biotypes, "data/gene_biotype_counts.tsv", "\t")
knitr::kable(biotypes[N_gene >= 100], digits=1)
```

# Gene counts


```{r}
trait <- trait[!is.na(trait$MAPPED_TRAIT_URI)]
trait$MAPPED_TRAIT <- iconv(trait$MAPPED_TRAIT, from="latin1", to="UTF-8")
sprintf("MAPPED_GENE values: %d", uniqueN(assn$MAPPED_GENE))
assn_reported <- assn[, .(STUDY_ACCESSION, `REPORTED_GENE(S)`)]
assn_reported <- unique(assn_reported[, list(GENE=unlist(strsplit(`REPORTED_GENE(S)`, ", *"))), by=STUDY_ACCESSION])
sprintf("REPORTED_GENE values: %d", uniqueN(assn_reported$GENE))
```

TCRD mapping.

```{r}
gsyms_tcrd <- unique(tcrd$tcrdGeneSymbol) #mapped
ensgs_tcrd <- unique(tcrd$ensemblGeneId) #mapped
sprintf("TCRD targets: %d ; ENSGs: %d; geneSymbols: %d", nrow(tcrd), length(ensgs_tcrd), length(gsyms_tcrd))
gsyms_tiga <- unique(snp2gene$GSYMB)
ensgs_tiga <- unique(ensbInfo$id)
gsyms_common <- intersect(gsyms_tiga, gsyms_tcrd)
ensgs_common <- intersect(ensgs_tiga, ensgs_tcrd)
sprintf("GSYMBs mapped to TCRD: %d", length(gsyms_common))
sprintf("ENSGs mapped to TCRD: %d", length(ensgs_common))
tcrd <- merge(tcrd, data.table(ensg=ensgs_tiga, in_gwascat=rep(T, length(ensgs_tiga))),
	by.x="ensemblGeneId", by.y="ensg", all.x=T, all.y=F)
tcrd$in_gwascat[is.na(tcrd$in_gwascat)] <- F
tcrd$TDL <- factor(tcrd$TDL, levels=c("NA", "Tdark", "Tbio", "Tchem", "Tclin"), ordered=T)
tdlcounts <- merge(tcrd[(in_gwascat), .(N_mapped_genes = .N), by="TDL"], tcrd[, .(N_genes = .N), by="TDL"], by="TDL")
tdlcounts <- rbindlist(list(tdlcounts, data.table(TDL=NA, N_mapped_genes=sum(tdlcounts[, N_mapped_genes]), N_genes=sum(tdlcounts[, N_genes]))))
knitr::kable(tdlcounts[order(-TDL)], caption="GWAS mapped gene counts by TDL")
```

# Gene-SNP-Study-Trait (G2T) associations

g2t should have one row for each gene-snp-study-trait association.

```{r}
g2t <- unique(snp2gene[, .(GSYMB, SNP, STUDY_ACCESSION)])
g2t <- merge(g2t, assn[, .(SNPS, STUDY_ACCESSION, PVALUE_MLOG, OR_or_BETA, oddsratio, beta)], 
	all.x=T, all.y=F, by.x=c("SNP", "STUDY_ACCESSION"), by.y=c("SNPS", "STUDY_ACCESSION"))

g2t <- merge(g2t, trait, all.x=F, all.y=F, by="STUDY_ACCESSION", allow.cartesian=T)
g2t <- g2t[!is.na(GSYMB)]
g2t <- g2t[!is.na(OR_or_BETA)]
g2t <- g2t[!grepl("(^LOC|^intergenic)", GSYMB)]

sprintf("GTs with pvalue_mlog, g2t: %d ; genes: %d ; traits: %d",
	 nrow(g2t[!is.na(g2t$PVALUE_MLOG),]),
	 uniqueN(g2t$GSYMB[!is.na(g2t$PVALUE_MLOG)]),
	 uniqueN(g2t$TRAIT[!is.na(g2t$PVALUE_MLOG)]))
sprintf("GTs with or_or_beta, g2t: %d ; genes: %d ; traits: %d",
	 nrow(g2t[!is.na(g2t$OR_or_BETA),]),
	 uniqueN(g2t$GSYMB[!is.na(g2t$OR_or_BETA)]),
	 uniqueN(g2t$TRAIT[!is.na(g2t$OR_or_BETA)]))
sprintf("GTs with oddsratio, g2t: %d ; genes: %d ; traits: %d",
	 nrow(g2t[!is.na(g2t$oddsratio),]),
	 uniqueN(g2t$GSYMB[!is.na(g2t$oddsratio)]),
	 uniqueN(g2t$TRAIT[!is.na(g2t$oddsratio)]))
sprintf("GTs with beta, g2t: %d ; genes: %d ; traits: %d",
	 nrow(g2t[!is.na(g2t$beta),]),
	 uniqueN(g2t$GSYMB[!is.na(g2t$beta)]),
	 uniqueN(g2t$TRAIT[!is.na(g2t$beta)]))
```

# GWAS traits and EFO

EFO = Experimental Factor Ontology. Includes GO, Orphanet, PO, Mondo and Uberon classes.
TSV from source OWL. 

```{r warning=FALSE}
efo <- read_delim(ifile_efo, "\t", col_types=cols(.default=col_character()))
setDT(efo)
efo[, source := sub("^.*/", "", source)] #Uri2Id
efo[, target := sub("^.*/", "", target)] #Uri2Id
efo_classes <- efo[node_or_edge=="node"][, `:=`(node_or_edge=NULL, source=NULL, target=NULL)]
efo_classes[, in_gwascat := as.logical(id %in% trait$efoId)]
efo_counts <- efo_classes[, Ontology := sub("_.*$", "", id)][, .(N_in_gwas = sum(in_gwascat), N_total = .N), by=Ontology][order(-N_in_gwas)]
sprintf("EFO total classes: %d; in GWAS: %d", uniqueN(efo_classes$id), sum(efo_counts$N_in_gwas))
knitr::kable(efo_counts[N_in_gwas>0], caption="EFO sources (GWAS-relevant)")
```

## GWAS trait-subclass relationships 

```{r}
efo_sub <- efo[node_or_edge=="edge"][, `:=`(node_or_edge=NULL, id=NULL, comment=NULL, uri=NULL)]
sprintf("EFO classes: %d ; total subclass relationships: %d", uniqueN(efo$id), nrow(efo_sub))
efo_sub[, source_in_gwas := (source %in% trait$efoId)]
efo_sub[, target_in_gwas := (target %in% trait$efoId)]
efo_sub <- efo_sub[(source_in_gwas) & (target_in_gwas)]
efo_sub[, `:=`(source_in_gwas=NULL, target_in_gwas=NULL, label=NULL)]
setnames(efo_sub, old=c("source", "target"), new=c("efoId", "subclass_id"))
trait_counts <- trait[, .(N_study = uniqueN(STUDY_ACCESSION)), by=c("efoId", "TRAIT")]
efo_sub <- merge(efo_sub, trait_counts[, .(efoId, trait_N_study=N_study, trait_name=TRAIT)], by="efoId", all.x=T, all.y=F, allow.cartesian=T)
efo_sub <- merge(efo_sub, trait_counts[, .(efoId, subclass_N_study=N_study, subclass_name=TRAIT)], by.x="subclass_id", by.y="efoId", all.x=T, all.y=F, allow.cartesian=T)
efo_sub <- efo_sub[, .(efoId, trait_name, subclass_id, subclass_name, trait_N_study, subclass_N_study)]
sprintf("GWAS trait-subclass pairs: %d", nrow(efo_sub))
setorder(efo_sub, -trait_N_study, -subclass_N_study)
knitr::kable(efo_sub[1:24], caption="Top EFO trait-subclass pairs")
```

## GWAS studies related by EFO subclass links.

```{r}
efo_gwas <- read_delim(ifile_efo_gwas, "\t", col_types=cols(.default=col_character()))
setDT(efo_gwas)
efo_gwas[, `:=`(efoId = sub("^.*/", "", trait_uri), subclass_id = sub("^.*/", "", subclass_uri))] #Uri2Id
sprintf("GWAS studies related by EFO-subclass: %d", 
        uniqueN(union(efo_gwas[study_accession != study_accession_subclass, study_accession], efo_gwas[study_accession != study_accession_subclass, study_accession_subclass])))
```

## EFO hierarchical groupings

One-to-many and many-to-one mappings exist.

```{r}
efo_groups <- read_delim(ifile_efo_groups, "\t", col_types=cols(.default=col_character(), in_gwc=col_logical(), level=col_integer(), N_sub=col_integer(), N_sub_gwc=col_integer()))
setDT(efo_groups)
setorder(efo_groups, -N_sub_gwc)
knitr::kable(efo_groups[(in_gwc)][1:24, .(Id, label, level, N_sub_gwc)], caption="GWAS traits with most sub-classes in dataset")
```

# GENE-TRAIT stats

Read gt_stats.tsv, built by tiga_gt_stats.R for TIGA. Statistics designed to weigh evidence aggregated across
studies, for each gene-trait association.


```{r}
gt <- read_delim(gt_statsfile, "\t", col_types=cols(.default=col_character(), n_study=col_integer(), study_N_mean=col_integer(), n_snp=col_integer(), geneNtrait=col_integer(), traitNgene=col_integer(), pvalue_mlog_max=col_double(), or_median=col_double(), n_beta=col_integer(), meanRank=col_double(), meanRankScore=col_double()))
setDT(gt)
sprintf("nrow(gt) = %d", nrow(gt))
```

### Multivariate non-parametric ranking via meanRankScore.

Since weighting among variables of merit is subjective, ranking by *meanRankScore* is
used. 

```{r}
knitr::kable(gt[trait=="pathological myopia"], caption="Example: pathological myopia")
```

### Focus on traits with more data and evidence, likely high scientific interest. But not too polygenic?

Which traits are most studied, based on N_study? But also, which of these may be too polygenic to
be suitably specific for drug discovery?

```{r}
xmax <- 10
trait_study_counts <- trait[, .(N_study = uniqueN(STUDY_ACCESSION)), by=c("MAPPED_TRAIT", "MAPPED_TRAIT_URI")]
trait_study_counts <- trait_study_counts[, .(trait_count = .N), by="N_study"][order(N_study)]
knitr::kable(rbindlist(list(trait_study_counts[1:xmax], data.table(N_study=sprintf(">%d", xmax), trait_count=trait_study_counts[N_study>xmax, sum(trait_count)]))), caption="Trait study-count distribution")
```

Most-studied traits

```{r}
trait_top <- trait[, .(MAPPED_TRAIT, N_study=uniqueN(STUDY_ACCESSION)), by="efoId"]
trait_top <- unique(merge(trait_top, gt[, .(N_genes=uniqueN(ensemblId)), by="efoId"], by="efoId", all.x=T, all.y=F))
#write_delim(trait_top[, .(efoId, MAPPED_TRAIT, N_study, N_genes)], "data/trait_counts.tsv", "\t")
knitr::kable(trait_top[order(-N_study), .(MAPPED_TRAIT, N_study, N_genes)][1:18], caption="Most-studied traits")
```

Most-polygenic traits

```{r}
knitr::kable(trait_top[order(-N_genes), .(MAPPED_TRAIT, N_study, N_genes)][1:28], caption="Most-polygenic traits")
```

# Prototype TIGA web app.

TIGA: Target Illumination GWAS Analytics

## Plot single trait with all associated genes.

* X-axis: Evidence (`meanRankScore`)
* Y-axis: Effect (`OR` or `N_beta`)
* Variables used in `meanRankScore`:
 * `pVal` - max of trait-SNP pValues
 * `N_snpw` - # SNPs implicating gene, weighted by genomic distance
 * `RCRAS` - Relative Citation Ratio Aggregated Score 
* Other variables:
 * `N_study` - # studies supporting association
 * `N_trait` - traits associated with gene
 * `N_snp` - # SNPs implicating gene
 * `OR` - median of trait-SNP ORs
 * `N_beta` - # supporting beta values
 * `N_gene` - # genes associated with trait
 * `N_sample` - mean of sample sizes


Color unmapped gray.

Plot for a selected trait:

```{r}
trait_id <- "MONDO_0005180" #Parkinson disease

query_trait <- gt[efoId==trait_id, first(trait)]
```

### Rank and filter based on meanRank score.

```{r}
t2g <- gt[efoId==trait_id]
t2g <- t2g[!is.na(ensemblId)]
t2g <- t2g[!is.na(or_median)]
setorder(t2g, meanRank)
NMAX_PLOT <- 100
t2g[, ok2plot := as.logical(.I <= NMAX_PLOT)]
message(sprintf("meanRankScore_filter result: %d / %d", sum(t2g[, ok2plot]), nrow(t2g)))
```

### Top hits:

```{r}
setorder(t2g, meanRank)
knitr::kable(t2g[1:10][, .(GSYMB=geneSymbol, name=geneName, family=geneFamily, geneIdgTdl, n_study, study_N_mean, rcras, or_median, pvalue_mlog_median, meanRankScore, meanRank)], caption=sprintf("TIGA: %s, top hits", query_trait))
```

### Plot

```{r}
t2g[is.na(geneIdgTdl), geneIdgTdl := "NA"]
t2g$geneIdgTdl <- factor(t2g$geneIdgTdl, levels=c("Tclin", "Tchem", "Tbio", "Tdark", "NA"), ordered=T)
#t2g[, markersize := 10*n_study]
t2g[, markersize := 10]
t2g[, text := paste0("<b>", geneSymbol, "</b> (", ensemblId, ")<br><b>", geneName, "</b><br>", "Family: ", geneFamily, "<br>", "TDL: ", geneIdgTdl, "<br>", 
"N_traits=", geneNtrait, " ; N_snp=", n_snp, " ; N_study=", n_study, " ; study_N = ", study_N_mean, "<br>", 
"OR = ", or_median, "; ",  study_N_mean, "; pVal=", sprintf("%.2g", 10^(-pvalue_mlog_max)), " ; RCRAS = ", rcras, ";<br>", 
"meanRankScore = ", sprintf("%.1f", meanRankScore), "; ",
"meanRank = ", sprintf("%.1f", meanRank), "; ")]
plot_ly(type='scatter', mode='markers', data=t2g[(ok2plot)], 
  color=~geneIdgTdl, colors=c("blue", "green", "red", "black", "gray"),
	x=~meanRankScore,
	y=~or_median, marker=list(symbol="circle", size=~markersize), text=~text) %>%
  layout(xaxis=list(title="Evidence (meanRankScore)", zeroline=F, showline=F), 
         yaxis=list(title="Effect (OddsRatio)"), 
        title=paste0("TIGA: ", query_trait, "<br>(", trait_id, ")"),
        margin=list(t=100,r=50,b=60,l=60), legend=list(orientation="h", x=1, y=1, xanchor="right"), showlegend=T, font=list(family="monospace", size=18)) %>%
  add_annotations(text=sprintf("N: %d (%d shown)", nrow(t2g), sum(t2g$ok2plot)), showarrow=F, x=0, y=1.0, xref="paper", yref="paper")
```



